lucene/core/src/java/org/apache/lucene/util/fst/FSTEnum.java

```diff
@@ -141,11 +141,65 @@ protected void doSeekCeil() throws IOException {
       //System.out.println("  cycle upto=" + upto + " arc.label=" + arc.label + " (" + (char) arc.label + ") vs targetLabel=" + targetLabel);
 
       if (arc.bytesPerArc != 0 && arc.label != -1) {
+        // Arcs are in an array
```

Comment:
I feel tempted to split those very large methods into smaller units (methods); it'd really make them clearer.

Reply:
I was thinking of introducing multiple Arc classes so that the (now three) different kinds of Arc decoding methods can be modeled explicitly with function overrides and the branching between different implementations done using virtual function dispatch rather than inline conditionals. That would likely subsume any refactoring here, so I'd like to postpone it, but I definitely agree with the temptation. Still, if you think it would make the review easier, I'd be happy to dive into that, but overall it will probably lead to more code changes making direct comparison difficult, and I was thinking to do a separate a refactoring commit with no behavior changes.

Reply:
After all, this revision includes break-up into smaller functions; it's definitely easier to read, and I guess could feed into some restructuring into classes later.

Reply:
Yes, a full refactoring is probably worth another issue, I'm just saying those methods evolved into complex beasts.